---
title: "Feed-microbiome-host interactions in Atlantic salmon over life stages"
subtitle: "Low dosage of betamannan"
author: "Shashank Gupta"
date: "2023-02-03"
output:
  html_document:
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
---


```{r, include=FALSE}
options(tinytex.verbose = TRUE)
```

```{r, include=FALSE}
options(knitr.duplicate.label = 'allow')

```

```{r setup, include=FALSE} 
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
```

## Step 1. Omics-Integration
### Load packages
```{r}
library("stringr")
library("dplyr")
library("DESeq2")
library("ggplot2")
library("tibble")
library("readr")
library("dplyr")
library("BiocParallel")
library("DT")
library("WGCNA")
library("cowplot")
library("gridExtra")
```


```{r}
take_average = T
cluster_the_columns = F
col_cut <- 1
row_cut <- 1
prefix_genes <- "h"
prefix_OTUs <- "m"
```

### Import omics data
```{r}
wgcna_metatranscriptomics <- readRDS("/Users/shashankgupta/Desktop/ImprovAFish/github/ImprovaFish/OmicsIntegration/stage2results_MetaTranscriptomics.rds")
wgcna_transcriptomics <- readRDS("/Users/shashankgupta/Desktop/ImprovAFish/github/ImprovaFish/OmicsIntegration/stage2results_Transcriptomics.rds")
wgcna_metagenomics <- readRDS("/Users/shashankgupta/Desktop/ImprovAFish/github/ImprovaFish/OmicsIntegration/stage2results_Metagenomics.rds")
wgcna_metabolomics_HILIC <- readRDS("/Users/shashankgupta/Desktop/ImprovAFish/github/ImprovaFish/OmicsIntegration/stage2results_HILIC.rds")
wgcna_metabolomics_RP <- readRDS("/Users/shashankgupta/Desktop/ImprovAFish/github/ImprovaFish/OmicsIntegration/stage2results_MetaTranscriptomics.rds")
wgcna_metabolomics_Lipidomics <- readRDS("/Users/shashankgupta/Desktop/ImprovAFish/github/ImprovaFish/OmicsIntegration/stage2results_Lipidomics.rds")
wgcna_metaprotomics <- readRDS("/Users/shashankgupta/Desktop/ImprovAFish/github/ImprovaFish/OmicsIntegration/stage2results_Proteomics.rds")
```



```{r}
# Function to remove and average modules
remove_and_average_MEs <- function(stage2results_X, take_average){
  idx <- which(colnames(stage2results_X$modules$MEs) == "ME0")
  stage2results_X$modules$MEs <- stage2results_X$modules$MEs[,-idx]
  
  if(take_average){
    stage2results_X$modules$MEs %>% 
      rownames_to_column(var = "common") %>%
      mutate(common = sub("_[0-9]{1,2}$", "", common)) %>% 
      group_by(common) %>% 
      summarise_all(list(mean)) %>% 
      ungroup() -> stage2results_X_eigengenes
    
    stage2results_X_eigengenes <- as.data.frame(stage2results_X_eigengenes)
    
  } else{
    stage2results_X_eigengenes <- stage2results_X$modules$MEs
  }
  
  rownames(stage2results_X_eigengenes) <- stage2results_X_eigengenes$common
  stage2results_X_eigengenes$common <- NULL
  
  return(stage2results_X_eigengenes)
}
```


```{r}
stage2results_X_eigengenes <- remove_and_average_MEs(stage2results_X = stage2results_X, take_average = T)



# Create a dendrogram of the transcriptomics eigengenes to organise the final plots.
X_ME_dendro <- hclust(as.dist(1 - WGCNA::bicor(stage2results_X_eigengenes, maxPOutliers = 0.05)), method = "ward.D2")

# "..palettes that vary from one hue to another via white may have a more symmetrical appearance in RGB space"
if(take_average){
  heatmap_colors <- colorRampPalette(c("#18b29f","#FFFFFF","#ac6721"), interpolate = "spline", space = "rgb")(51)
} else{
  heatmap_colors <- colorRampPalette(c("#18b29f","#FFFFFF","#ac6721"), interpolate = "spline", space = "rgb")(51)
}


annotation_col <- 
  rownames(stage2results_X_eigengenes) %>% 
  as.data.frame() %>% 
  dplyr::rename(common = ".") %>% 
  tidyr::separate(col = common,  into = c("Time", "New_Diet"), remove = F, extra = "drop") %>% # If replicates are present, we drop them here
  dplyr::select(common,"Time", "New_Diet") %>%  
  tibble::column_to_rownames(var = "common")
colnames(annotation_col) <- c("Time" , "New_Diet")
annotation_colors <- list(
  Time = c(`T1` = "navy", `T2` = "darkgreen", `T3` = "red"),
  New_Diet = c(`ctr` = "#1D88AF", `mc1` = "darkblue", `mc2` = "darkgreen", `mn3` = "#F08A46"))


stage2results_X_eigengenes_to_plot <- 
  dplyr::inner_join(annotation_col %>% 
                      rownames_to_column(var = "common"), 
                    stage2results_X_eigengenes %>% 
                      rownames_to_column(var = "common"), 
                    by = "common") %>%
  #dplyr::arrange(Feed, Water, Day) %>%              # The order at which the columns should appear, given that there is no clustering.
  dplyr::select(common, starts_with("ME")) %>% 
  tibble::column_to_rownames(var = "common") %>% 
  t()
library(WGCNA)
moduleLabels = stage2results_X$modules$colors
moduleColors = labels2colors(stage2results_X$modules$colors)
table(moduleLabels)
table(moduleColors)
table(moduleLabels, moduleColors)


#heatmap_colors <- colorRampPalette(rev(RColorBrewer::brewer.pal(n = 6, name ="RdBu")))(51)
#For Expression data heatmap color
heatmap_colors <- colorRampPalette(c("#18b29f","#FFFFFF","#ac6721"), interpolate = "spline", space = "rgb")(51)

pheatmap::pheatmap(stage2results_X_eigengenes_to_plot, 
                   cluster_cols = F,
                   cluster_rows = X_ME_dendro,
                   treeheight_row = 20,
                   #cluster_rows = T,
                   cutree_rows = row_cut,
                   cutree_cols = col_cut,
                   color = heatmap_colors,
                   fontsize = 10,
                   fontsize_col = ifelse(take_average, 10, 6),
                   fontsize_row = 8,
                   annotation_colors = annotation_colors,
                   annotation_col = annotation_col, 
                   show_colnames = F,
                   silent = F,annotation_legend = F,
                   labels_row = paste0(prefix_genes, rownames(stage2results_X_eigengenes_to_plot)),
                   main = paste("Gene Module 'expression'\n", ifelse(take_average, "mean values", ""))) -> X_plot





#stage2results_Y_backup <- stage2results_Y
#stage2results_Y <- stage2results_Y_backup
stage2results_Y   
# Remove ME0 from analysis
stage2results_Y_eigengenes <- remove_and_average_MEs(stage2results_X = stage2results_Y, take_average = T)

dim(stage2results_Y_eigengenes); dim(stage2results_X_eigengenes)


#Correlate modules from transcriptomics and metagenomics.
# Check that the samples are in the same order. 
# If they are not in order, change their order to match; If they do not match one-to-one, call an error.
same_order <- all(rownames(stage2results_Y_eigengenes) == rownames(stage2results_X_eigengenes))
if(!same_order){
  stage2results_Y_eigengenes <- stage2results_Y_eigengenes[order(rownames(stage2results_Y_eigengenes)),]
  stage2results_X_eigengenes <- stage2results_X_eigengenes[order(rownames(stage2results_X_eigengenes)),]
  same_order <- all(rownames(stage2results_Y_eigengenes) == rownames(stage2results_X_eigengenes))
  if(!same_order){
    stop("Sample names do not match. Samples should be identical.", call. = F)
  }
} else{cat("Samples match")}
 

#####
p.value_matr <- corr.value_matr <- matrix(ncol = ncol(stage2results_Y_eigengenes), 
                                          nrow = ncol(stage2results_X_eigengenes), 
                                          dimnames = list(colnames(stage2results_X_eigengenes), 
                                                          colnames(stage2results_Y_eigengenes)))


for(i in 1:ncol(stage2results_X_eigengenes)){
  for(j in 1:ncol(stage2results_Y_eigengenes)){
    cor.res <- cor.test(stage2results_X_eigengenes[,i], stage2results_Y_eigengenes[,j])
    p.value_matr[i, j] <- cor.res$p.value
    corr.value_matr[i, j] <- cor.res$estimate
  }
}

# Correct for number of tests
p.value_matr.adjust <- p.adjust(p.value_matr, method = "fdr")
dim(p.value_matr.adjust) <- dim(p.value_matr)
dimnames(p.value_matr.adjust) <- list(colnames(stage2results_X_eigengenes), colnames(stage2results_Y_eigengenes))


# Add significance level.  
# One star means a p-value of less than 0.05; Two stars is less than 0.01, and three, is less than 0.001.

signif_matrix <- rep("", length(p.value_matr))
three_star <- which( p.value_matr <= 0.001)
signif_matrix[three_star] <- "***"
two_star <- which((p.value_matr <= 0.01) & (p.value_matr > 0.001))
signif_matrix[two_star] <- "**"
one_star <- which((p.value_matr <= 0.05) & (p.value_matr > 0.01))
signif_matrix[one_star] <- "*"
dim(signif_matrix) = dim(p.value_matr) # Give textMatrix the correct dimensions 


# Collect all results into a list.
Y_corr_X <- list(p_value = p.value_matr, 
                 p_value_adj = p.value_matr.adjust,
                 signif_matrix = signif_matrix,
                 correlation = corr.value_matr)
rm(p.value_matr, p.value_matr.adjust, signif_matrix, corr.value_matr)



heatmap_colors <- colorRampPalette(rev(RColorBrewer::brewer.pal(n = 6, name ="RdBu")))(51)


stage2results_Y$hubs %>% 
  as.data.frame() %>% 
  dplyr::rename("OTU_name" = ".") %>%
  tibble::rownames_to_column(var = "Module") -> hubOTUs

library(phyloseq)
otu_taxonomy <- data.frame(tax_table(all))
dplyr::left_join(hubOTUs, 
                 (otu_taxonomy %>%
                    tibble::rownames_to_column(var = "OTU_name")), 
                 by = "OTU_name") -> hubOTUs_Before

hubOTUs_Before$ModuleOTU <- paste0("ME" ,hubOTUs_Before$Module)


test <- Y_corr_X$correlation
hubOTUs_Before <- hubOTUs_Before[, c(10, 8)]
#hubOTUs_Before$Genus <- replace(hubOTUs_Before$Genus, hubOTUs_Before$Genus=="Allorhizobium_Neorhizobium_Pararhizobium_Rhizobium", "Rhizobium")
#hubOTUs_Before$Genus <- replace(hubOTUs_Before$Genus, hubOTUs_Before$Genus=="Burkholderia_Caballeronia_Paraburkholderia", "Paraburkholderia")
hubOTUs_Before$Genus <- replace(hubOTUs_Before$Genus, hubOTUs_Before$Genus=="BBMC_4", "Family_Fibrobacterales")
hubOTUs_Before$Genus <- replace(hubOTUs_Before$Genus, hubOTUs_Before$Genus=="Prochlorococcus_MIT9313", "Prochlorococcus")


match(hubOTUs_Before[, "ModuleOTU"], colnames(test))
colnames(test)[match(hubOTUs_Before[,"ModuleOTU"], colnames(test))] = hubOTUs_Before[,"Genus"]

Y_corr_X$correlation <- test

pheatmap::pheatmap(Y_corr_X$correlation, 
                   color = heatmap_colors, 
                   treeheight_col = 0, 
                   treeheight_row = 0,  # will be shown on the transcriptomics ME heatmap
                   cluster_rows = X_ME_dendro,
                   cutree_rows = row_cut,
                   display_numbers = Y_corr_X$signif_matrix, 
                   fontsize_number = 8, #10
                   breaks = seq(from = -1, to = 1, length.out = 51), 
                   silent = F,
                   show_rownames = F, legend = F,
                   labels_row = paste0(prefix_OTUs, rownames(Y_corr_X$correlation)),
                   labels_col = paste0(colnames(Y_corr_X$correlation)),
                   main = "Eigen ASVs") -> Y_corr_X_plot

out = binarizeCategoricalVariable(Z_traits,
                                  includePairwise = FALSE,
                                  includeLevelVsAll = TRUE)


#Z_traits_backup <- Z_traits
#Correlate transcriptomics with traits

sample_info <- as_tibble(sample_info)
Z_traits <- sample_info %>% dplyr::select(ID_New, Time, New_Diet)
Z_traits <- Z_traits %>%
  mutate(Time = factor(Time, levels = c("T1", "T2", "T3")),
         New_Diet = factor(New_Diet),
         Day = ifelse(Time == "T1", paste0("T1_", New_Diet),
                      ifelse(Time == "T2", paste0("T2_", New_Diet), paste0("T3_", New_Diet))))
Z_traits$Time <- factor(Z_traits$Time, levels = c("T1", "T2", "T3"))
Z_traits$Day <- factor(Z_traits$Day)

# One-hot encoding the nominal variables
nominal_variables <- Z_traits %>%
  dplyr::select(one_of("Day", "New_Diet", "Time" )) %>% 
  model.matrix(~.-1, data = .) %>% 
  as_tibble()

ordinal_and_other_variables <- Z_traits %>% 
  dplyr::select(one_of("Day", "New_Diet", "Time" , "ID_New"))
Z_traits <- nominal_variables
row.names(Z_traits) <- ordinal_and_other_variables$ID_New
Z_traits$ID_New <- row.names(Z_traits) 
Z_traits$TimeT1 <- ifelse(grepl("T1", Z_traits$ID_New), 1, 0)
Z_traits$New_Dietctr <- ifelse(grepl("ctr", Z_traits$ID_New), 1, 0)
row.names(Z_traits)  <- Z_traits$ID_New
take_average <- TRUE
if (take_average) {
  Z_traits <- Z_traits %>% 
    mutate(ID_New = sub("_[0-9]{1,2}$", "", ID_New)) %>% 
    group_by(ID_New) %>% 
    summarise_all(list(mean), na.rm = TRUE) %>%  # Remove NA values
    ungroup() %>% 
    tibble::column_to_rownames(var = "ID_New")
  Z_traits <- as.data.frame(Z_traits)
}

Z_traits <- Z_traits[, c(13:15, 18, 16, 17)]
colnames(Z_traits)[1:6] <- c("Diet_MC1", "Diet_MC2", "Diet_MN3", "Time_T1",  "Time_T2", "Time_T3")

# Check that the samples are in the same order. 
# If they are not in order, change their order to match; If they do not match one-to-one, call an error.
same_order <- all(rownames(Z_traits) == rownames(stage2results_X_eigengenes))
if(!same_order){
  stage2results_Y_eigengenes <- stage2results_Y_eigengenes[order(rownames(stage2results_Y_eigengenes)),]
  stage2results_X_eigengenes <- stage2results_X_eigengenes[order(rownames(stage2results_X_eigengenes)),]
  same_order <- all(rownames(stage2results_Y_eigengenes) == rownames(stage2results_X_eigengenes))
  if(!same_order){
    stop("Sample names do not match. Samples should be identical.", call. = F)
  }
} else{cat("Samples match")}

p.value_matr <- corr.value_matr <- matrix(ncol = ncol(Z_traits), 
                                          nrow = ncol(stage2results_X_eigengenes), 
                                          dimnames = list(colnames(stage2results_X_eigengenes), 
                                                          colnames(Z_traits)))
for(i in 1:ncol(stage2results_X_eigengenes)){
  for(j in 1:ncol(Z_traits)){
    
    if(colnames(Z_traits)[j] == "Diet_MC1"){
      cor.res <- cor.test(stage2results_X_eigengenes[,i], Z_traits[,j], method = "spearman", exact = FALSE)
      p.value_matr[i, j] <- cor.res$p.value
      corr.value_matr[i, j] <- cor.res$estimate
    } else{
      cor.res <- cor.test(stage2results_X_eigengenes[,i], Z_traits[,j])
      p.value_matr[i, j] <- cor.res$p.value
      corr.value_matr[i, j] <- cor.res$estimate
    }
  }
}


# Correct for number of tests
p.value_matr.adjust <- p.adjust(p.value_matr, method = "fdr")
dim(p.value_matr.adjust) <- dim(p.value_matr)
dimnames(p.value_matr.adjust) <- list(colnames(stage2results_X_eigengenes), colnames(Z_traits))


# Add significance level.  
# One star means a p-value of less than 0.05; Two stars is less than 0.01, and three, is less than 0.001.

signif_matrix <- rep("", length(p.value_matr))
three_star <- which( p.value_matr <= 0.001)
signif_matrix[three_star] <- "***"
two_star <- which((p.value_matr <= 0.01) & (p.value_matr > 0.001))
signif_matrix[two_star] <- "**"
one_star <- which((p.value_matr <= 0.05) & (p.value_matr > 0.01))
signif_matrix[one_star] <- "*"
dim(signif_matrix) = dim(p.value_matr) # Give textMatrix the correct dimensions 


# Collect all results into a list.
Z_corr_X <- list(p_value = p.value_matr, 
                 p_value_adj = p.value_matr.adjust,
                 signif_matrix = signif_matrix,
                 correlation = corr.value_matr)
rm(p.value_matr, p.value_matr.adjust, signif_matrix, corr.value_matr)


heatmap_colors <- colorRampPalette(rev(RColorBrewer::brewer.pal(n = 6, name ="RdBu")))(51)

pheatmap::pheatmap(Z_corr_X$correlation, 
                   color = heatmap_colors,
                   cluster_cols = F,
                   treeheight_col = 0, 
                   treeheight_row = 0, # will be shown on the transcriptomics ME heatmap
                   cluster_rows = X_ME_dendro,
                   cutree_rows = row_cut,
                   display_numbers = Z_corr_X$signif_matrix, 
                   fontsize_number = 8, # 10
                   breaks = seq(from = -1, to = 1, length.out = 51), 
                   silent = F, 
                   legend = F,
                   show_rownames = F) -> Z_corr_X_plot


# Modules with no significant associations
if (filter.sig.mod == 1) {
  idx1 <- rowSums(#Z_corr_X$signif_matrix == "*" |
    Z_corr_X$signif_matrix == "**" |
      Z_corr_X$signif_matrix == "***") > 0
  idx2 <- rowSums(#Y_corr_X$signif_matrix == "*" |
    Y_corr_X$signif_matrix == "**" | 
      Y_corr_X$signif_matrix == "***") > 0
  sig.modules.h <- rownames(Y_corr_X$p_value)[idx1 & idx2]
  
  sig.modules.m <- colnames(Y_corr_X$p_value)[colSums(#Y_corr_X$signif_matrix[idx1 & idx2,] == "*" |
    Y_corr_X$signif_matrix[idx1 & idx2,] == "**" | 
      Y_corr_X$signif_matrix[idx1 & idx2,] == "***") > 0]
  
  
  save(sig.modules.h, sig.modules.m, file = "../data/module_data/sig_modules.Rdata")
}

cowplot::plot_grid(Z_corr_X_plot$gtable,
                   X_plot$gtable,
                   Y_corr_X_plot$gtable, 
                   ncol = 3, 
                   rel_widths = c(dim(Z_traits)[2]/3, 
                                  ifelse(take_average,dim(stage2results_X_eigengenes)[1]/2,dim(stage2results_X_eigengenes)[1]/6),
                                  dim(stage2results_Y_eigengenes)[2]/2.2),
                   align = "h") + ggplot2::theme(plot.margin = ggplot2::unit(c(3,0,2.5,1), "cm")) # Top, Right, Bottom, Left






########################################################################
########################################################################
# moduleâ€“trait associations

# Define numbers of genes and samples
#omics_data_host <- t(omics_data_host)
nGenes = ncol(omics_data_host);
nSamples = nrow(omics_data_host);
moduleLabels1 = stage2results_X$modules$colors
moduleColors1 = labels2colors(stage2results_X$modules$colors)
MEs = stage2results_X$modules$MEs

table(moduleColors1)
table(moduleLabels1)


# Recalculate MEs with color labels
MEs0 = moduleEigengenes(omics_data_host, moduleLabels1)$eigengenes
#MEs = orderMEs(MEs0)
head(MEs0)
MEs <- MEs0[, c(2:25)]
head(MEs)
#colnames(MEs)<- paste("h", colnames(MEs), sep = "")
#host_gut_mapping_file <- read.table("../host_data/host_gut_mapping.txt", sep = "\t", row.names = 1, header = T)

setwd("/Users/shashankgupta/Desktop/ImprovAFish/ImprovAFish")
sample_info1 <- read.csv("sample_info.csv", row.names = 1)
sample_info1[1:7] <- NULL
sample_info1$ID_New <- NULL
sample_info1$common <- NULL

reorder_idx <- match(row.names(MEs), row.names(sample_info1))
sample_info1 <- sample_info1[ reorder_idx , ]
all(row.names(sample_info1) %in% row.names(MEs))
all(row.names(sample_info1) == row.names(MEs))

moduleTraitCor = cor(MEs, sample_info1, use = "p")
moduleTraitCor <- moduleTraitCor[order(match(rownames(moduleTraitCor), c("ME4","ME1","ME24","ME7","ME15","ME9","ME22","ME6","ME19","ME8","ME20","ME10","ME16","ME18","ME13","ME12","ME5","ME11","ME3","ME14","ME21","ME23","ME17","ME2"))), ]


moduleTraitPvalue = corPvalueStudent(moduleTraitCor, nSamples)

# Will display correlations and their p-values
textMatrix =  paste(signif(moduleTraitCor, 2), "\n(",
                    signif(moduleTraitPvalue, 1), ")", sep = "");
dim(textMatrix) = dim(moduleTraitCor)
#par(mar = c(6, 8, 1, 1))
# Display the correlation values within a heatmap plot
labeledHeatmap(Matrix = moduleTraitCor,
               xLabels = names(sample_info1),
               yLabels = rownames(moduleTraitCor),
               ySymbols =rownames(moduleTraitCor),
               colorLabels = FALSE,
               colors = blueWhiteRed(50),
               textMatrix = textMatrix,
               setStdMargins = FALSE,
               cex.text = 0.5,
               zlim = c(-1,1),
               main = paste("Module-trait relationships"))



signif_matrix <- rep("", length(moduleTraitPvalue))
three_star <- which( moduleTraitPvalue <= 0.001)
signif_matrix[three_star] <- "***"
two_star <- which((moduleTraitPvalue <= 0.01) & (moduleTraitPvalue > 0.001))
signif_matrix[two_star] <- "**"
one_star <- which((moduleTraitPvalue <= 0.05) & (moduleTraitPvalue > 0.01))
signif_matrix[one_star] <- "*"
dim(signif_matrix) = dim(moduleTraitPvalue) # Give textMatrix the correct dimensions 


pheatmap::pheatmap(moduleTraitCor, 
                   color = heatmap_colors,
                   yLabels = names(MEs),
                   ySymbols = names(MEs),
                   cluster_cols = F,
                   #treeheight_col = 0, 
                   #treeheight_row = 0, # will be shown on the transcriptomics ME heatmap
                   cluster_rows = F,
                   #cutree_rows = row_cut,
                   display_numbers = signif_matrix, 
                   fontsize_number = 8, # 10
                   breaks = seq(from = -1, to = 1, length.out = 51), 
                   silent = F, 
                   legend = F,
                   show_rownames = F,
                   main = "Traits and external variables") -> MT_plot






cowplot::plot_grid(MT_plot$gtable,
                   Z_corr_X_plot$gtable,
                   X_plot$gtable,
                   MetaTrans_corr_X_plot$gtable,
                   Y_corr_X_plot$gtable,
                   Proteomics_corr_X_plot$gtable,
                   Metabolites_1_corr_X_plot$gtable,
                   Metabolites_2_corr_X_plot$gtable,
                   Lipidomics_corr_X_plot$gtable,
                   ncol = 9,
                   align = 'h',
                   rel_widths = c(1.4, 0.6, 4.5, 0.8, 1.3,1.3 , 0.3, 0.3, 0.6) # PNG 1700 (Width) * 1000 (Height)  # pdf 18 * 9 portrait
)+ ggplot2::theme(plot.margin = ggplot2::unit(c(5,2,3,2), "cm"))






##############################################################################################################
##############################################################################################################
# From Meta-MetaTranscriptomics
#stage2results_MetaTrans_backup <- stage2results_MetaTrans
# Remove ME0 from analysis
idx <- which(colnames(stage2results_MetaTrans$modules$MEs) == "ME0")
stage2results_MetaTrans$modules$MEs <- stage2results_MetaTrans$modules$MEs[,-idx]

take_average <- TRUE
if(take_average){
  stage2results_MetaTrans$modules$MEs %>% 
    rownames_to_column(var = "common") %>% 
    mutate(common = sub("_HGm[0-9]{1,2}$", "", common)) %>% 
    group_by(common) %>% 
    summarise_all(list(mean)) %>% 
    ungroup() -> stage2results_MetaTrans_eigengenes
  
  stage2results_MetaTrans_eigengenes <- as.data.frame(stage2results_MetaTrans_eigengenes)
} else{
  stage2results_MetaTrans_eigengenes <- stage2results_MetaTrans$modules$MEs
}

rownames(stage2results_MetaTrans_eigengenes) <- stage2results_MetaTrans_eigengenes$common
stage2results_MetaTrans_eigengenes$common <- NULL

dim(stage2results_MetaTrans_eigengenes); dim(stage2results_X_eigengenes)



table(rownames(stage2results_X_eigengenes) %in% rownames(stage2results_MetaTrans_eigengenes))

stage2results_X_eigengenes_New <- stage2results_X_eigengenes[rownames(stage2results_X_eigengenes) %in% rownames(stage2results_MetaTrans_eigengenes), ]
table(rownames(stage2results_X_eigengenes_New) %in% rownames(stage2results_MetaTrans_eigengenes))

#Correlate modules from transcriptomics and metagenomics.
# Check that the samples are in the same order. 
# If they are not in order, change their order to match; If they do not match one-to-one, call an error.
same_order <- all(rownames(stage2results_MetaTrans_eigengenes) == rownames(stage2results_X_eigengenes_New))
if(!same_order){
  stage2results_MetaTrans_eigengenes <- stage2results_MetaTrans_eigengenes[order(rownames(stage2results_MetaTrans_eigengenes)),]
  stage2results_X_eigengenes_New <- stage2results_X_eigengenes_New[order(rownames(stage2results_X_eigengenes_New)),]
  same_order <- all(rownames(stage2results_MetaTrans_eigengenes) == rownames(stage2results_X_eigengenes_New))
  if(!same_order){
    stop("Sample names do not match. Samples should be identical.", call. = F)
  }
} else{cat("Samples match")}


#####
p.value_matr <- corr.value_matr <- matrix(ncol = ncol(stage2results_MetaTrans_eigengenes), 
                                          nrow = ncol(stage2results_X_eigengenes_New), 
                                          dimnames = list(colnames(stage2results_X_eigengenes_New), 
                                                          colnames(stage2results_MetaTrans_eigengenes)))


for(i in 1:ncol(stage2results_X_eigengenes_New)){
  for(j in 1:ncol(stage2results_MetaTrans_eigengenes)){
    cor.res <- cor.test(stage2results_X_eigengenes_New[,i], stage2results_MetaTrans_eigengenes[,j])
    p.value_matr[i, j] <- cor.res$p.value
    corr.value_matr[i, j] <- cor.res$estimate
  }
}

# Correct for number of tests
p.value_matr.adjust <- p.adjust(p.value_matr, method = "fdr")
dim(p.value_matr.adjust) <- dim(p.value_matr)
dimnames(p.value_matr.adjust) <- list(colnames(stage2results_X_eigengenes_New), colnames(stage2results_MetaTrans_eigengenes))


# Add significance level.  
# One star means a p-value of less than 0.05; Two stars is less than 0.01, and three, is less than 0.001.

signif_matrix <- rep("", length(p.value_matr))
three_star <- which( p.value_matr <= 0.001)
signif_matrix[three_star] <- "***"
two_star <- which((p.value_matr <= 0.01) & (p.value_matr > 0.001))
signif_matrix[two_star] <- "**"
one_star <- which((p.value_matr <= 0.05) & (p.value_matr > 0.01))
signif_matrix[one_star] <- "*"
dim(signif_matrix) = dim(p.value_matr) # Give textMatrix the correct dimensions 


# Collect all results into a list.
MetaTrans_corr_X <- list(p_value = p.value_matr, 
                         p_value_adj = p.value_matr.adjust,
                         signif_matrix = signif_matrix,
                         correlation = corr.value_matr)
rm(p.value_matr, p.value_matr.adjust, signif_matrix, corr.value_matr)

heatmap_colors <- colorRampPalette(rev(RColorBrewer::brewer.pal(n = 6, name ="RdBu")))(51)

#Annotation for Eigen Metatranscriptomics
meta_info <- read_tsv("CountsKallistoQuantificationDeNovoSTxDRAM/Total.DRAM.STx.bac.k2.total.annotations.tsv")
rownames(meta_info) <- meta_info$...1
stage2results_MetaTrans$hubs %>% 
  as.data.frame() %>% 
  dplyr::rename("metatranscriptomics_name" = ".") %>%
  tibble::rownames_to_column(var = "Module") -> hubmetatranscriptomics


dplyr::left_join(hubmetatranscriptomics, 
                 (meta_info %>%
                    tibble::rownames_to_column(var = "metatranscriptomics_name")), 
                 by = "metatranscriptomics_name") -> hubmetatranscriptomics_Before

hubmetatranscriptomics_Before$Modulemetatranscriptomic <- paste0("ME" ,hubmetatranscriptomics_Before$Module)
hubmetatranscriptomics_Before <- hubmetatranscriptomics_Before[, c(33, 15)]

#Match with Eigen Metatranscriptomics
test <- MetaTrans_corr_X$correlation
match(hubmetatranscriptomics_Before[, "Modulemetatranscriptomic"], colnames(test))
colnames(test)[match(hubmetatranscriptomics_Before[,"Modulemetatranscriptomic"], colnames(test))] = hubmetatranscriptomics_Before[,"uniref_taxonomy"]

MetaTrans_corr_X$correlation <- test

pheatmap::pheatmap(MetaTrans_corr_X$correlation, 
                   color = heatmap_colors, 
                   treeheight_col = 0, 
                   treeheight_row = 0,  # will be shown on the transcriptomics ME heatmap
                   cluster_rows = X_ME_dendro,
                   cutree_rows = row_cut,
                   display_numbers = MetaTrans_corr_X$signif_matrix, 
                   fontsize_number = 8, #10
                   breaks = seq(from = -1, to = 1, length.out = 51), 
                   silent = F,
                   show_rownames = F, legend = F,
                   labels_row = paste0(prefix_OTUs, rownames(MetaTrans_corr_X$correlation)),
                   labels_col = paste0(colnames(MetaTrans_corr_X$correlation)),
                   main = paste("Eigen\n", "Metatranscriptomics" )) -> MetaTrans_corr_X_plot

##











##############################################################################################################
##############################################################################################################
# From Meta-Proteomics
#stage2results_Proteomics_backup <- stage2results_Proteomics
#rownames(stage2results_Proteomics$modules$MEs) <- gsub("-MC1-TK1-", "_MC1_tk1_", rownames(stage2results_Proteomics$modules$MEs))
# Remove ME0 from analysis
idx <- which(colnames(stage2results_Proteomics$modules$MEs) == "ME0")
stage2results_Proteomics$modules$MEs <- stage2results_Proteomics$modules$MEs[,-idx]

take_average<- TRUE
if(take_average){
  stage2results_Proteomics$modules$MEs %>% 
    rownames_to_column(var = "common") %>% 
    mutate(common = sub("_HGm.*", "", common)) %>% 
    group_by(common) %>% 
    summarise_all(list(mean)) %>% 
    ungroup() -> stage2results_Proteomics_eigengenes
  
  stage2results_Proteomics_eigengenes <- as.data.frame(stage2results_Proteomics_eigengenes)
} else{
  stage2results_Proteomics_eigengenes <- stage2results_Proteomics$modules$MEs
}

rownames(stage2results_Proteomics_eigengenes) <- stage2results_Proteomics_eigengenes$common
stage2results_Proteomics_eigengenes$common <- NULL

dim(stage2results_Proteomics_eigengenes); dim(stage2results_X_eigengenes)



table(rownames(stage2results_X_eigengenes) %in% rownames(stage2results_Proteomics_eigengenes))

stage2results_X_eigengenes_New <- stage2results_X_eigengenes[rownames(stage2results_X_eigengenes) %in% rownames(stage2results_Proteomics_eigengenes), ]


table( rownames(stage2results_X_eigengenes_New) %in% rownames(stage2results_Proteomics_eigengenes))
table( rownames(stage2results_X_eigengenes_New) == rownames(stage2results_Proteomics_eigengenes))

#Correlate modules from transcriptomics and metagenomics.
# Check that the samples are in the same order. 
# If they are not in order, change their order to match; If they do not match one-to-one, call an error.
same_order <- all(rownames(stage2results_Proteomics_eigengenes) == rownames(stage2results_X_eigengenes_New))
if(!same_order){
  stage2results_Proteomics_eigengenes <- stage2results_Proteomics_eigengenes[order(rownames(stage2results_Proteomics_eigengenes)),]
  stage2results_X_eigengenes_New <- stage2results_X_eigengenes_New[order(rownames(stage2results_X_eigengenes_New)),]
  same_order <- all(rownames(stage2results_Proteomics_eigengenes) == rownames(stage2results_X_eigengenes_New))
  if(!same_order){
    stop("Sample names do not match. Samples should be identical.", call. = F)
  }
} else{cat("Samples match")}


#####
p.value_matr <- corr.value_matr <- matrix(ncol = ncol(stage2results_Proteomics_eigengenes), 
                                          nrow = ncol(stage2results_X_eigengenes_New), 
                                          dimnames = list(colnames(stage2results_X_eigengenes_New), 
                                                          colnames(stage2results_Proteomics_eigengenes)))


for(i in 1:ncol(stage2results_X_eigengenes_New)){
  for(j in 1:ncol(stage2results_Proteomics_eigengenes)){
    cor.res <- cor.test(stage2results_X_eigengenes_New[,i], stage2results_Proteomics_eigengenes[,j])
    p.value_matr[i, j] <- cor.res$p.value
    corr.value_matr[i, j] <- cor.res$estimate
  }
}

# Correct for number of tests
p.value_matr.adjust <- p.adjust(p.value_matr, method = "fdr")
dim(p.value_matr.adjust) <- dim(p.value_matr)
dimnames(p.value_matr.adjust) <- list(colnames(stage2results_X_eigengenes_New), colnames(stage2results_Proteomics_eigengenes))


# Add significance level.  
# One star means a p-value of less than 0.05; Two stars is less than 0.01, and three, is less than 0.001.

signif_matrix <- rep("", length(p.value_matr))
three_star <- which( p.value_matr <= 0.001)
signif_matrix[three_star] <- "***"
two_star <- which((p.value_matr <= 0.01) & (p.value_matr > 0.001))
signif_matrix[two_star] <- "**"
one_star <- which((p.value_matr <= 0.05) & (p.value_matr > 0.01))
signif_matrix[one_star] <- "*"
dim(signif_matrix) = dim(p.value_matr) # Give textMatrix the correct dimensions 


# Collect all results into a list.
Proteomics_corr_X <- list(p_value = p.value_matr, 
                          p_value_adj = p.value_matr.adjust,
                          signif_matrix = signif_matrix,
                          correlation = corr.value_matr)
rm(p.value_matr, p.value_matr.adjust, signif_matrix, corr.value_matr)



heatmap_colors <- colorRampPalette(rev(RColorBrewer::brewer.pal(n = 6, name ="RdBu")))(51)


stage2results_Proteomics$hubs %>% 
  as.data.frame() %>% 
  dplyr::rename("Proteomics_name" = ".") %>%
  tibble::rownames_to_column(var = "Module") -> hubProteomics

hubProteomics -> hubProteomics_Before

hubProteomics_Before$ModuleProteomics <- paste0("ME" ,hubProteomics_Before$Module)
test <- Proteomics_corr_X$correlation
hubProteomics_Before$Proteomics_name <- sub("XP_.*", "", hubProteomics_Before$Proteomics_name)
hubProteomics_Before$Proteomics_name <- sub("NP_.*", "", hubProteomics_Before$Proteomics_name)
hubProteomics_Before$Proteomics_name <- replace(hubProteomics_Before$Proteomics_name, hubProteomics_Before$Proteomics_name=="cytochrome.c.oxidase.subunit.5B..mitochondrial", "Cytochrome c oxidase subunit 5B, mitochondrial")
hubProteomics_Before$Proteomics_name <- replace(hubProteomics_Before$Proteomics_name, hubProteomics_Before$Proteomics_name=="myosin.7.like.isoform.X1", "myosin-7 isoform X1")
hubProteomics_Before$Proteomics_name <- replace(hubProteomics_Before$Proteomics_name, hubProteomics_Before$Proteomics_name=="creatine.kinase.S.type..mitochondrial", "Creatine kinase S-type, mitochondrial")
hubProteomics_Before$Proteomics_name <- replace(hubProteomics_Before$Proteomics_name, hubProteomics_Before$Proteomics_name=="X3.hydroxyisobutyrate.dehydrogenase..mitochondrial", "3-hydroxyisobutyrate dehydrogenase, mitochondrial")
hubProteomics_Before$Proteomics_name <- replace(hubProteomics_Before$Proteomics_name, hubProteomics_Before$Proteomics_name=="LOW.QUALITY.PROTEIN..zonadhesin.like", "Zonadhesin like")
hubProteomics_Before$Proteomics_name <- replace(hubProteomics_Before$Proteomics_name, hubProteomics_Before$Proteomics_name=="cytosolic.non.specific.dipeptidase", "Cytosol nonspecific dipeptidas")
hubProteomics_Before$Proteomics_name <- replace(hubProteomics_Before$Proteomics_name, hubProteomics_Before$Proteomics_name=="protocadherin.Fat.4", "Protocadherin Fat 4")
hubProteomics_Before$Proteomics_name <- replace(hubProteomics_Before$Proteomics_name, hubProteomics_Before$Proteomics_name=="cadherin.17.precursor", "Cadherins 17 precursor")
hubProteomics_Before$Proteomics_name <- replace(hubProteomics_Before$Proteomics_name, hubProteomics_Before$Proteomics_name=="sushi.domain.containing.protein.2.isoform.X2", "Sushi domain-containing protein 2")
hubProteomics_Before$Proteomics_name <- replace(hubProteomics_Before$Proteomics_name, hubProteomics_Before$Proteomics_name=="intestinal.type.alkaline.phosphatase", "Intestinal-type alkaline phosphatase")

match(hubProteomics_Before[, "ModuleProteomics"], colnames(test))
colnames(test)[match(hubProteomics_Before[,"ModuleProteomics"], colnames(test))] = hubProteomics_Before[,"Proteomics_name"]

Proteomics_corr_X$correlation <- test
pheatmap::pheatmap(Proteomics_corr_X$correlation, 
                   color = heatmap_colors, 
                   treeheight_col = 0, 
                   treeheight_row = 0,  # will be shown on the transcriptomics ME heatmap
                   cluster_rows = X_ME_dendro,
                   cutree_rows = row_cut,
                   display_numbers = Proteomics_corr_X$signif_matrix, 
                   fontsize_number = 8, #10
                   breaks = seq(from = -1, to = 1, length.out = 51), 
                   silent = F,
                   show_rownames = F,
                   legend = F,
                   labels_row = paste0(prefix_OTUs, rownames(Proteomics_corr_X$correlation)),
                   labels_col = paste0(colnames(Proteomics_corr_X$correlation)),
                   main = paste("Eigen\n", "Meta-proteins" )
) -> Proteomics_corr_X_plot


```{r, warning=FALSE, message=FALSE, eval=FALSE}
MEs = moduleEigengenes(Proteomics_host, moduleLabels1)$eigengenes

MEs <- orderMEs(MEs)
module_order = names(MEs) %>% gsub("ME","", .)
MEs0 <- MEs
MEs0$Sample.ID = row.names(MEs)

mME = MEs0 %>%
  pivot_longer(-Sample.ID) %>%
  mutate(
    name = gsub("ME", "", name),
    name = factor(name, levels = module_order)
  )



library(tidyverse)
setwd("/Users/shashankgupta/Desktop/ImprovAFish/ImprovAFish")
sample_info3 <- read.csv("sample_info.csv", row.names = 1)
sample_info3$Sample.ID <- sample_info3$ID_New
mME_meta <- merge(mME, sample_info3, by = "Sample.ID") 
colnames(mME_meta)[2] <-"Module"

mME_meta$New_Diet <- factor(mME_meta$New_Diet, levels = c("CTR", "MC1", "MC2", "MN3"))
mME_meta$Module <- factor(mME_meta$Module, levels = c("0", "1", "2", "3", "4", "5", "6", "7", "8", "9",
                                                      "10", "11", "12", "13", "14", "15", "16", "17",
                                                      "18", "19", "20","21","22","23", "24"))
expression_plots<-mME_meta%>%
  #  group_by(Module) %>%
  ggplot(aes(x=Time, y=value, fill=New_Diet)) +
  facet_wrap(~ Module)+
  ylab("Mean Module Eigenegene Value") +
  geom_hline(yintercept = 0, linetype="dashed", color = "grey")+
  geom_boxplot(width=.5, outlier.shape= NA, position = position_dodge(width = 0.5), alpha = 0.7) +
  stat_summary(fun=mean, geom="line", aes(New_Diet=New_Diet, color = New_Diet), position = position_dodge(width = 0.5))  + 
  geom_point(pch = 21, position = position_dodge(width = 0.5)) +
  scale_fill_manual(name="Lifestage", values=c("#8C510A", "#DFC27D","#80CDC1", "#003C30", "#BA55D3")) +
  scale_color_manual(name="Lifestage", values=c("#8C510A", "#DFC27D","#80CDC1", "#003C30", "#BA55D3")) + 
  #xlab("Hours Post-Fertilization") + #Axis titles
  theme_bw() + theme(panel.border = element_rect(color="black", fill=NA, size=0.75), panel.grid.major = element_blank(), #Makes background theme white
                     panel.grid.minor = element_blank(), axis.line = element_blank()) +
  theme(axis.text = element_text(size = 11 , color = "black"),
        axis.title = element_text(size = 16, color = "black"), 
        axis.text.x = element_text(size=11, color="black"), 
        legend.title=element_blank(), 
        legend.text=element_text(color="black", size=12)); expression_plots


```








################################################################################################
################################################################################################
##
# From Metabolomics_1 HILIC

#stage2results_HILIC_backup <- stage2results_HILIC
stage2results_HILIC
rownames(stage2results_HILIC$modules$MEs) <- gsub("-MC1-TK1-", "_mc1_tk1_", rownames(stage2results_HILIC$modules$MEs))

# Remove ME0 from analysis
idx <- which(colnames(stage2results_HILIC$modules$MEs) == "ME0")
stage2results_HILIC$modules$MEs <- stage2results_HILIC$modules$MEs[,-idx]


if(take_average){
  stage2results_HILIC$modules$MEs %>% 
    rownames_to_column(var = "common") %>% 
    mutate(common = sub("_VFA.*", "", common)) %>% 
    group_by(common) %>% 
    summarise_all(list(mean)) %>% 
    ungroup() -> stage2results_HILIC_eigengenes
  
  stage2results_HILIC_eigengenes <- as.data.frame(stage2results_HILIC_eigengenes)
} else{
  stage2results_HILIC_eigengenes <- stage2results_HILIC$modules$MEs
}

rownames(stage2results_HILIC_eigengenes) <- stage2results_HILIC_eigengenes$common
stage2results_HILIC_eigengenes$common <- NULL

dim(stage2results_HILIC_eigengenes); dim(stage2results_X_eigengenes)
table(rownames(stage2results_X_eigengenes) %in% rownames(stage2results_HILIC_eigengenes))
stage2results_X_eigengenes_New <- stage2results_X_eigengenes[rownames(stage2results_X_eigengenes) %in% rownames(stage2results_HILIC_eigengenes), ]
table( rownames(stage2results_X_eigengenes_New) %in% rownames(stage2results_HILIC_eigengenes))

#Correlate modules from transcriptomics and metagenomics.

# Check that the samples are in the same order. 
# If they are not in order, change their order to match; If they do not match one-to-one, call an error.
same_order <- all(rownames(stage2results_HILIC_eigengenes) == rownames(stage2results_X_eigengenes_New))
if(!same_order){
  stage2results_HILIC_eigengenes <- stage2results_HILIC_eigengenes[order(rownames(stage2results_HILIC_eigengenes)),]
  stage2results_X_eigengenes_New <- stage2results_X_eigengenes_New[order(rownames(stage2results_X_eigengenes_New)),]
  same_order <- all(rownames(stage2results_HILIC_eigengenes) == rownames(stage2results_X_eigengenes_New))
  if(!same_order){
    stop("Sample names do not match. Samples should be identical.", call. = F)
  }
} else{cat("Samples match")}


#####
p.value_matr <- corr.value_matr <- matrix(ncol = ncol(stage2results_HILIC_eigengenes), 
                                          nrow = ncol(stage2results_X_eigengenes_New), 
                                          dimnames = list(colnames(stage2results_X_eigengenes_New), 
                                                          colnames(stage2results_HILIC_eigengenes)))


for(i in 1:ncol(stage2results_X_eigengenes_New)){
  for(j in 1:ncol(stage2results_HILIC_eigengenes)){
    cor.res <- cor.test(stage2results_X_eigengenes_New[,i], stage2results_HILIC_eigengenes[,j])
    p.value_matr[i, j] <- cor.res$p.value
    corr.value_matr[i, j] <- cor.res$estimate
  }
}

# Correct for number of tests
p.value_matr.adjust <- p.adjust(p.value_matr, method = "fdr")
dim(p.value_matr.adjust) <- dim(p.value_matr)
dimnames(p.value_matr.adjust) <- list(colnames(stage2results_X_eigengenes_New), colnames(stage2results_HILIC_eigengenes))


# Add significance level.  
# One star means a p-value of less than 0.05; Two stars is less than 0.01, and three, is less than 0.001.

signif_matrix <- rep("", length(p.value_matr))
three_star <- which( p.value_matr <= 0.001)
signif_matrix[three_star] <- "***"
two_star <- which((p.value_matr <= 0.01) & (p.value_matr > 0.001))
signif_matrix[two_star] <- "**"
one_star <- which((p.value_matr <= 0.05) & (p.value_matr > 0.01))
signif_matrix[one_star] <- "*"
dim(signif_matrix) = dim(p.value_matr) # Give textMatrix the correct dimensions 


# Collect all results into a list.
Metabolites_1_corr_X <- list(p_value = p.value_matr, 
                             p_value_adj = p.value_matr.adjust,
                             signif_matrix = signif_matrix,
                             correlation = corr.value_matr)
rm(p.value_matr, p.value_matr.adjust, signif_matrix, corr.value_matr)



heatmap_colors <- colorRampPalette(rev(RColorBrewer::brewer.pal(n = 6, name ="RdBu")))(51)


stage2results_HILIC$hubs %>% 
  as.data.frame() %>% 
  dplyr::rename("metabolites_HILIC_name" = ".") %>%
  tibble::rownames_to_column(var = "Module") -> hubmetabolites_HILIC

hubmetabolites_HILIC -> hubmetabolites_HILIC_Before

hubmetabolites_HILIC_Before$Modulemetabolites_HILIC <- paste0("ME" ,hubmetabolites_HILIC_Before$Module)


test <- Metabolites_1_corr_X$correlation
#hubmetabolites_HILIC_Before$metabolites_HILIC_name <- replace(hubmetabolites_HILIC_Before$metabolites_HILIC_name, hubmetabolites_HILIC_Before$metabolites_HILIC_name=="Ile.Leu...244.17825.Da.264.70.s", "Ile-Leu")
#hubmetabolites_HILIC_Before$metabolites_HILIC_name <- replace(hubmetabolites_HILIC_Before$metabolites_HILIC_name, hubmetabolites_HILIC_Before$metabolites_HILIC_name=="Di.N.octyl.phthalate...390.27623.Da.707.55.s", "Di-N-octylphthalate")
hubmetabolites_HILIC_Before$metabolites_HILIC_name <- replace(hubmetabolites_HILIC_Before$metabolites_HILIC_name, hubmetabolites_HILIC_Before$metabolites_HILIC_name=="Asp.Asp_248.06507Da665.09s", "Asp-Asp")
hubmetabolites_HILIC_Before$metabolites_HILIC_name <- replace(hubmetabolites_HILIC_Before$metabolites_HILIC_name, hubmetabolites_HILIC_Before$metabolites_HILIC_name=="Glu.Thr_248.10146Da623.48s", "Glu-Thr")
#colnames(test)[1:2] <- c("D-Arabinonic acid", "Thr-Thr")

match(hubmetabolites_HILIC_Before[, "Modulemetabolites_HILIC"], colnames(test))
colnames(test)[match(hubmetabolites_HILIC_Before[,"Modulemetabolites_HILIC"], colnames(test))] = hubmetabolites_HILIC_Before[,"metabolites_HILIC_name"]

Metabolites_1_corr_X$correlation <- test

pheatmap::pheatmap(Metabolites_1_corr_X$correlation, 
                   color = heatmap_colors, 
                   treeheight_col = 0, 
                   treeheight_row = 0,  # will be shown on the transcriptomics ME heatmap
                   cluster_rows = X_ME_dendro,
                   cutree_rows = row_cut,
                   display_numbers = Metabolites_1_corr_X$signif_matrix, 
                   fontsize_number = 8, #10
                   breaks = seq(from = -1, to = 1, length.out = 51), 
                   silent = F,
                   legend = F,
                   show_rownames = F,
                   labels_row = paste0(prefix_OTUs, rownames(Metabolites_1_corr_X$correlation)),
                   labels_col = paste0(colnames(Metabolites_1_corr_X$correlation)),
                   #main = "Eigen metabolites_HILIC"
) -> Metabolites_1_corr_X_plot

```










